// @elara-services/sdk@2.0.2/dist/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.routes = exports.links = exports.status = exports.userAgent = void 0;
const package_json_1 = require("../package.json");
const base = `elara.workers.dev`;
exports.userAgent = `${package_json_1.name} (${package_json_1.version}, https://github.com/elara-bots/npm)`;
function status(message) {
    const obj = {
        status: false,
        message: `Unknown error while trying to get the data.`,
    };
    if (message instanceof Error) {
        obj.message = message.message;
    }
    else if (typeof message === "string") {
        obj.message = message;
    }
    return obj;
}
exports.status = status;
exports.links = {
    docs: `https://elara-services-sdk.pages.dev`,
    base: `https://services.${base}`,
    haste: `https://h.${base}`,
    support: `https://services.${base}/support`,
};
exports.routes = {
    ping: "/site/ping",
    bin: {
        get: (id) => `/bin/api/${id}`,
        post: `/bin/api`,
    },
    dbl: {
        get: (token, id) => `/api/dbl/stats?id=${id}&token=${token}`,
        post: (token, id, shards, servers) => `/api/dbl/post?id=${id}&servers=${servers}&shards=${shards}&token=${token}`,
    },
    api: {
        lyrics: (name) => `/api/lyrics?name=${name}`,
        photos: (image) => `/api/photos/${image}`,
        special: (image) => `/api/special?type=${image}`,
        math: `/api/math`,
        translate: `/api/translate`,
        facts: (type) => `/api/facts?type=${type.toLowerCase()}`,
        memes: (clean) => `/api/photos/memes?clean=${clean}`,
        ball: `/api/8ball`,
        npm: (name) => `/api/npm?name=${name}`,
        time: {
            place: (name) => `/api/time?place=${name.toString().toLowerCase()}`,
            all: `/api/time?all=true`,
        },
        docs: (search, project, branch) => `/api/discord.js-docs?search=${search}&project=${project}&branch=${branch}`,
    },
    automod: {
        images: (token, percent) => `/api/automod/images?token=${token}&percent=${percent}`,
        words: `/api/automod/words`,
        links: `/api/automod/links`,
    },
    dev: {
        blacklist: {
            servers: {
                toggle: ({ id, action, name, reason, mod, }) => `/dev/blacklists/servers?id=${id}&action=${action}&name=${encodeURIComponent(name)}&reason=${encodeURIComponent(reason)}&mod=${encodeURIComponent(mod)}`,
                list: (id) => `/dev/blacklists/servers?id=${id}`,
            },
            users: {
                toggle: ({ id, action, username, tag, reason, mod, }) => `/dev/blacklists/users?id=${id}&action=${action}&username=${encodeURIComponent(username)}&tag=${encodeURIComponent(tag)}&reason=${encodeURIComponent(reason)}&mod=${encodeURIComponent(mod)}`,
                list: (id) => `/dev/blacklists/users?id=${id}`,
            },
        },
    },
    platform: {
        ytstats: (user, token) => `/api/platform/yt-stats?user=${user}&token=${token}`,
        roblox: (id, group) => `/api/platform/roblox${group ? `-group` : ""}?id=${id}`,
        fortnite: (user, token, platform) => `/api/platform/fortnite?user=${user}&token=${token}&platform=${platform}`,
        imdb: (name, token) => `/api/platform/imdb?token=${token}&show=${name}`,
        ytsearch: (name, type, token) => `/api/platform/yt-search?token=${token}&name=${name}&type=${type}`,
        picarto: (search) => `/api/platform/picarto?search=${search}`,
    },
};
