// @elara-services/webhooks@2.1.18/dist/src/webhooks/discord/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.url = exports.split = exports.status = exports.error = exports.validateURL = exports.defaultOptions = void 0;
const package_json_1 = require("../../../package.json");
exports.defaultOptions = {
    username: "",
    avatar_url: "",
    threadId: ""
};
function validateURL(url) {
    if (!url || typeof url !== "string")
        return false;
    if (!url.match(/https?:\/\/(www.|canary.|ptb.)?discord(app)?.com\/api\/|https?:\/\/services.superchiefyt.workers.dev/gi))
        return false;
    return true;
}
exports.validateURL = validateURL;
;
function error(e) {
    throw new Error(`[${package_json_1.name}, ${package_json_1.version}]: ${e}`);
}
exports.error = error;
function status(status, data) {
    return { status, data };
}
exports.status = status;
function split(url) {
    let [id, token] = url.split("webhooks/")[1].split("/") || [null, null];
    if (!id || !token)
        return null;
    return { id, token };
}
exports.split = split;
function url(url) {
    try {
        let Url = new URL(url);
        return {
            path: Url.pathname.split("api/")[1],
            query: Url.search,
            thread_id: Url.searchParams.get("thread_id") || undefined
        };
    }
    catch (e) {
        return null;
    }
}
exports.url = url;
