// @elara-services/utils@1.7.0/dist/src/lib/is.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.make = exports.noop = exports.makeArray = exports.getPluralTxt = exports.is = void 0;
const v10_1 = require("discord-api-types/v10");
exports.is = {
    string: (name, checkEmpty = true) => {
        if (typeof name === "string") {
            if (checkEmpty && !name) {
                return false;
            }
            return true;
        }
        return false;
    },
    number: (num, checkEmpty = true) => {
        if (typeof num === "number" && !isNaN(num)) {
            if (checkEmpty && num <= 0) {
                return false;
            }
            return true;
        }
        return false;
    },
    boolean: (bool) => {
        return typeof bool === "boolean";
    },
    array: (arr, checkEmpty = true) => {
        if (Array.isArray(arr)) {
            if (checkEmpty && !arr.length) {
                return false;
            }
            return true;
        }
        return false;
    },
    object: (obj) => {
        return typeof obj === "object";
    },
    error: (err) => {
        return err instanceof Error;
    },
    undefined: (und) => {
        return typeof und === "undefined" || und === undefined;
    },
    null: (name) => {
        return name === null;
    },
    promise: (promise) => {
        return promise instanceof Promise;
    },
};
function getPluralTxt(arrOrNum) {
    if (exports.is.array(arrOrNum)) {
        if (arrOrNum.length >= 2) {
            return "s";
        }
    }
    else if (exports.is.number(arrOrNum)) {
        if (arrOrNum >= 2) {
            return "s";
        }
    }
    return "";
}
exports.getPluralTxt = getPluralTxt;
/**
 * @deprecated
 * @description Use make.array()
 */
function makeArray(arr) {
    return exports.make.array(arr);
}
exports.makeArray = makeArray;
function noop() {
    return null;
}
exports.noop = noop;
function getExtension(str, ex = "png", forceStatic = false) {
    if (str?.startsWith("a_") && !forceStatic) {
        ex = "gif";
    }
    return ex;
}
const main = `discord.com`;
exports.make = {
    array: (arr) => (exports.is.array(arr) ? arr : []),
    set: (val) => (val instanceof Set ? val : new Set()),
    cdn: (ex, proxy = false) => `https://${proxy ? `media.discordapp.net` : `cdn.discordapp.com`}/${ex}`,
    emojiURL: (id, ext = "png", size) => `${exports.make.cdn(v10_1.CDNRoutes.emoji(id, getExtension(id, ext, false)))}${size ? `?size=${size}` : ""}`,
    guildURL: (guildId) => `https://${main}/channels/${guildId}`,
    channelURL: (guildId, channelId) => `https://${main}/channels/${guildId}/${channelId}`,
    messageURL: (guildId, channelId, messageId) => `https://${main}/channels/${guildId}/${channelId}/${messageId}`,
    webhook: (id, token, options) => {
        const url = new URL(`https://${options?.build ? `${options.build}.` : ""}${main}/api${options?.version ? `/${options.version}` : ""}/webhooks/${id}/${token}${options?.message ? `/messages/${options?.message}` : ""}`);
        if (exports.is.boolean(options?.wait)) {
            url.searchParams.append(`wait`, `${options?.wait || "false"}`);
        }
        else if (exports.is.undefined(options?.wait)) {
            url.searchParams.append(`wait`, "true");
        }
        if (exports.is.string(options?.threadId)) {
            url.searchParams.append("thread_id", options?.threadId);
        }
        return url.toString();
    },
    image: {
        users: {
            avatar: (userId, avatar, forceStatic = false, ex = "png", size, proxy) => {
                if (!avatar) {
                    // @ts-ignore
                    const index = (userId >> 22) % 6;
                    return `${exports.make.cdn(v10_1.CDNRoutes.defaultUserAvatar(index), proxy)}${size ? `?size=${size}` : ""}`;
                }
                return `${exports.make.cdn(v10_1.CDNRoutes.userAvatar(userId, avatar, getExtension(avatar, ex, forceStatic)), proxy)}${size ? `?size=${size}` : ""}`;
            },
            banner: (userId, banner, forceStatic = false, ex = "png", size, proxy) => {
                return banner && `${exports.make.cdn(v10_1.CDNRoutes.userBanner(userId, banner, getExtension(banner, ex, forceStatic)), proxy)}${size ? `?size=${size}` : ""}`;
            },
        },
        guild: {
            icon: (guildId, icon, forceStatic = false, ex = "png", size, proxy) => {
                return icon && `${exports.make.cdn(v10_1.CDNRoutes.guildIcon(guildId, icon, getExtension(icon, ex, forceStatic)), proxy)}${size ? `?size=${size}` : ""}`;
            },
            banner: (guildId, banner, forceStatic = false, ex = "png", size, proxy) => {
                return banner && `${exports.make.cdn(v10_1.CDNRoutes.guildBanner(guildId, banner, getExtension(banner, ex, forceStatic)), proxy)}${size ? `?size=${size}` : ""}`;
            },
        },
        sticker: (id, ext = "png", size, proxy) => {
            return exports.make.cdn(`stickers/${id}.${ext}${size && ext !== "json" ? `?size=${size}` : ""}`, proxy);
        },
    },
};
