// @elara-services/utils@1.7.0/dist/src/lib/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.p = exports.parser = void 0;
const times_1 = require("./times");
async function parser(obj = {}, options = {}) {
    let str = JSON.stringify(obj);
    const r = (names, s) => {
        if (names.some((c) => c.includes("_"))) {
            names = [...names, ...names.map((c) => c.replace(/_/gi, ".")), ...names.map((c) => c.replace(/_/gi, "-")), ...names.map((c) => c.replace(/_/gi, ";")), ...names.map((c) => c.replace(/_/gi, ":"))];
        }
        str = str.replace(new RegExp(names.join("|"), "gi"), s);
    };
    if (options.member) {
        options.user = options.member.user;
        options.guild = options.member.guild;
    }
    if (options.guild) {
        const { guild } = options;
        const owner = guild.client.users.resolve(guild.ownerId) || (await guild.client.users.fetch(guild.ownerId, { cache: true }).catch(() => null));
        if (owner) {
            r([exports.p.owner.username], owner.username);
            r([exports.p.owner.mention, exports.p.owner.user], owner.toString());
            r([exports.p.owner.id], owner.id);
            r([exports.p.owner.avatar], owner.displayAvatarURL());
        }
        r([exports.p.guild.name, exports.p.guild.server], guild.name);
        r([exports.p.guild.icon], guild.iconURL());
        r([exports.p.guild.banner], (guild.banner ? guild.bannerURL() : ""));
        r([exports.p.guild.splash], guild.discoverySplashURL());
        r([exports.p.guild.created], guild.createdAt.toLocaleString());
        r([exports.p.guild.createdDiscordTimestamp], (0, times_1.timeFormat)(guild.createdAt, true, "f"));
        r([exports.p.guild.vanity], guild.vanityURLCode || "N/A");
        r([exports.p.guild.features], guild.features.join(", "));
    }
    if (options.user) {
        const { user } = options;
        r([exports.p.user.username], user.username);
        r([exports.p.user.mention, exports.p.user.user], user.toString());
        r([exports.p.user.id], user.id);
        r([exports.p.user.avatar], user.displayAvatarURL());
        r([exports.p.user.accentColor], user.hexAccentColor || "N/A");
        r([exports.p.user.banner], user.banner ? user.bannerURL() || "" : "");
        r([exports.p.user.created], user.createdAt.toLocaleString());
        r([exports.p.user.createdDiscordTimestamp], (0, times_1.timeFormat)(user.createdAt, true, "f"));
    }
    if (options.member) {
        const { member } = options;
        r([exports.p.member.avatar], member.displayAvatarURL());
        r([exports.p.member.joined], member.joinedAt?.toLocaleString() || "N/A");
        if (member.joinedAt) {
            r([exports.p.member.joinedDiscordTimestamp], (0, times_1.timeFormat)(member.joinedAt, true, "f"));
        }
        r([exports.p.member.nickname], member.nickname || "N/A");
        r([exports.p.member.roles], member.roles.cache.map((c) => `\`${c.name}\``).join(", "));
        if (member.communicationDisabledUntil) {
            r([exports.p.member.timeout], member.communicationDisabledUntil?.toLocaleString() || "N/A");
            r([exports.p.member.timeoutDiscordTimestamp], (0, times_1.timeFormat)(member.communicationDisabledUntil, true, "f"));
        }
    }
    return JSON.parse(str);
}
exports.parser = parser;
exports.p = {
    owner: {
        user: "%owner%",
        mention: "%owner_mention%",
        username: "%owner_username%",
        id: "%owner_id%",
        avatar: "%owner_avatar%",
    },
    guild: {
        server: "%server%",
        name: "%server_name%",
        icon: "%server_icon%",
        splash: "%server_splash%",
        banner: "%server_banner%",
        created: "%server_created%",
        createdDiscordTimestamp: "%server_created_timestamp%",
        vanity: "%server_vanity%",
        features: "%server_features%",
    },
    user: {
        user: "%user%",
        mention: "%user_mention%",
        username: "%user_username%",
        id: "%user_id%",
        avatar: "%user_avatar%",
        accentColor: "%user_accentColor%",
        banner: "%user_banner%",
        created: "%user_created%",
        createdDiscordTimestamp: "%user_created_timestamp%",
    },
    member: {
        joined: "%member_joined%",
        joinedDiscordTimestamp: "%member_joined_timestamp%",
        nickname: "%member_nickname%",
        roles: "%member_roles%",
        avatar: "%member_avatar%",
        timeout: "%member_timeout%",
        timeoutDiscordTimestamp: "%member_timeout_timestamp%",
    },
};
