// @elara-services/utils@1.7.0/dist/src/lib/times.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.time = exports.timeFormat = exports.getTimeRemaining = exports.getTimeLeft = exports.ms = exports.getTimeFrom = exports.error = exports.log = exports.times = void 0;
const builders_1 = require("@discordjs/builders");
const moment_1 = __importDefault(require("moment"));
const ms_1 = __importDefault(require("ms"));
// eslint-disable-next-line @typescript-eslint/no-var-requires
require("moment-duration-format")(moment_1.default);
// eslint-disable-next-line prefer-const
exports.times = {
    timeZone: "",
    short: "",
};
function log(...args) {
    if (!exports.times.timeZone) {
        return console.log(...args);
    }
    return console.log(`[${new Date().toLocaleString("en-US", { timeZone: exports.times.timeZone })}]: `, ...args);
}
exports.log = log;
function error(...args) {
    if (!exports.times.timeZone) {
        return console.error(...args);
    }
    return console.error(`[${new Date().toLocaleString("en-US", { timeZone: exports.times.timeZone })}]: `, ...args);
}
exports.error = error;
function getTimeFrom(amount, unit, date = new Date()) {
    return (0, moment_1.default)(new Date(date)).add(amount, unit).toISOString();
}
exports.getTimeFrom = getTimeFrom;
exports.ms = {
    get(ms, long = true) {
        return (0, ms_1.default)(ms, { long });
    },
    convert(seconds) {
        if (typeof seconds !== "number") {
            return "Removed";
        }
        if (seconds === 0) {
            return "Off";
        }
        const days = Math.floor(seconds / (24 * 60 * 60));
        seconds -= days * (24 * 60 * 60);
        const hours = Math.floor(seconds / (60 * 60));
        seconds -= hours * (60 * 60);
        const minutes = Math.floor(seconds / 60);
        seconds -= minutes * 60;
        return `${0 < days ? days + ` Day${days === 1 ? "" : "s"}, ` : ""}${hours === 0 ? "" : `${hours} Hour${hours === 1 ? "" : "s"}${(minutes && seconds) === 0 ? "" : ", "}`}${minutes !== 0 ? `${minutes} Minute${minutes === 1 ? "" : "s"}${seconds === 0 ? "" : ", "}` : ""}${seconds !== 0 ? `${Math.floor(seconds)} Second${seconds === 1 ? "" : "s"}` : ""}`;
    },
};
function getTimeLeft(date, type) {
    return (moment_1.default
        .duration(new Date(date).getTime() - new Date().getTime())
        // @ts-ignore
        .format(type)
        .toString()
        .startsWith("-"));
}
exports.getTimeLeft = getTimeLeft;
function getTimeRemaining(date, type, reverse = false) {
    return (moment_1.default
        .duration(reverse ? new Date(date).getTime() - new Date().getTime() : new Date().getTime() - new Date(date).getTime())
        // @ts-ignore
        .format(type));
}
exports.getTimeRemaining = getTimeRemaining;
function timeFormat(date, discordFormat = false, format = "f") {
    if (discordFormat) {
        return (0, builders_1.time)(new Date(date), format);
    }
    return `${new Date(date || new Date()).toLocaleString("en-US", exports.times.timeZone ? { timeZone: exports.times.timeZone } : undefined)}${exports.times.short ? ` (${exports.times.short})` : ""}`;
}
exports.timeFormat = timeFormat;
exports.time = {
    countdown: (ms) => timeFormat(new Date(Date.now() + ms), true, "R"),
    relative: (date) => timeFormat(date, true, "R"),
    long: {
        dateTime: (date) => timeFormat(date, true, "F"),
        time: (date) => timeFormat(date, true, "T"),
        date: (date) => timeFormat(date, true, "D"),
    },
    short: {
        dateTime: (date) => timeFormat(date, true, "f"),
        time: (date) => timeFormat(date, true, "t"),
        date: (date) => timeFormat(date, true, "d"),
    },
};
