// @elara-services/utils@1.7.0/dist/src/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchAttachment = exports.limits = exports.snowflakes = exports.getClientIdFromToken = exports.field = exports.commands = void 0;
const packages_1 = require("@elara-services/packages");
const snowflake_1 = require("@sapphire/snowflake");
const is_1 = require("./is");
const status_1 = require("./status");
function commands(content, prefix) {
    if (!is_1.is.string(content)) {
        return {
            name: "",
            args: [],
            hasPrefix() {
                return false;
            },
            isCommand() {
                return false;
            },
        };
    }
    const str = content?.split(/ +/g) || "";
    const name = str[0].slice(prefix.length).toLowerCase();
    return {
        name,
        args: str.slice(1),
        hasPrefix() {
            if (content.toLowerCase().startsWith(prefix)) {
                return true;
            }
            return false;
        },
        isCommand(commandName) {
            if (commandName === name) {
                return true;
            }
            return false;
        },
    };
}
exports.commands = commands;
function field(name = "\u200b", value = "\u200b", inline = false) {
    return { name, value, inline };
}
exports.field = field;
function getClientIdFromToken(token) {
    return Buffer.from(token.split(".")[0], "base64").toString();
}
exports.getClientIdFromToken = getClientIdFromToken;
exports.snowflakes = {
    get: (id) => {
        const data = snowflake_1.DiscordSnowflake.deconstruct(id);
        data.date = new Date(snowflake_1.DiscordSnowflake.timestampFrom(data.id));
        return data;
    },
    generate: (options) => snowflake_1.DiscordSnowflake.generate(options).toString(),
};
exports.limits = {
    audit: 512,
    content: 2000,
    embeds: {
        max: 10,
        characters: 6000,
    },
    title: 256,
    description: 4096,
    fields: 25,
    field: {
        name: 256,
        value: 1024,
    },
    footer: {
        text: 2048,
    },
    author: {
        name: 256,
    },
};
async function fetchAttachment(url) {
    if (!url.toLowerCase().includes(".json")) {
        return status_1.status.error(`You didn't provide a valid .json file.`);
    }
    const r = await (0, packages_1.fetch)(url);
    if (!r) {
        return status_1.status.error(`Unable to fetch the attachment data.`);
    }
    return status_1.status.data(r);
}
exports.fetchAttachment = fetchAttachment;
