// @elara-services/webhooks@2.1.18/dist/src/shared.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = exports._debug = exports.getComponents = exports.createWebhook = exports.rest = exports.disabledLogging = exports.debugLogging = exports.buttonIds = exports.channels = exports.webhooks = void 0;
const collection_1 = require("@discordjs/collection");
const rest_1 = require("@discordjs/rest");
const v10_1 = require("discord-api-types/v10");
const _1 = require(".");
const package_json_1 = require("../package.json");
exports.webhooks = new collection_1.Collection();
exports.channels = new collection_1.Collection();
exports.buttonIds = null;
exports.debugLogging = false;
exports.disabledLogging = [];
exports.rest = new rest_1.REST();
function createWebhook(client, id = "", name) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!id) {
            return null;
        }
        try {
            const res = (yield client.post(v10_1.Routes.channelWebhooks(id), {
                reason: "Created for logging purposes",
                body: { name }
            }));
            if (res.id && res.token) {
                exports.webhooks.set(id, { id: res.id, token: res.token });
                return res;
            }
            return null;
        }
        catch (err) {
            (0, _1.log)(`[WEBHOOK:CREATE:ERROR]: `, err);
            return null;
        }
    });
}
exports.createWebhook = createWebhook;
function getComponents(components, shouldTransform = true) {
    if (!Array.isArray(components)) {
        return;
    }
    if (!components.length) {
        return;
    }
    if (components.length >= 2 && exports.buttonIds && shouldTransform) {
        let list = components.flatMap((c) => c.components);
        // @ts-ignore
        components = (0, exports.buttonIds)([...new Set(list.filter(c => [2, "Button"].includes(c.type)).map(c => { var _a; return (c.custom_id || ((_a = c.data) === null || _a === void 0 ? void 0 : _a.customId) || c.customId || "").split(":")[1]; }))].map(c => ({ id: c })), true);
    }
    return components;
}
exports.getComponents = getComponents;
function _debug(...args) {
    if (!exports.debugLogging) {
        return;
    }
    console.log(`[${new Date().toLocaleString()}]: [${package_json_1.name}:${package_json_1.version}:DEBUGGER]: `, ...args);
}
exports._debug = _debug;
exports.setup = {
    debug: (toggle) => {
        exports.debugLogging = toggle;
    },
    buttonIds: (func) => {
        exports.buttonIds = func;
    },
    ignoreGuildId: (id) => {
        if (exports.disabledLogging.includes(id)) {
            exports.disabledLogging = exports.disabledLogging.filter(c => c !== id);
            return `Removed (${id}) from the 'disabledLogging' array.`;
        }
        exports.disabledLogging.push(id);
        return `Added (${id}) to the 'disabledLogging' array.`;
    }
};
