// @elara-services/packages@6.1.0/dist/src/lib/discord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashBuilder = exports.Duration = exports.Interactions = exports.ButtonStyle = void 0;
const utils_1 = require("@elara-services/utils");
const v10_1 = require("discord-api-types/v10");
exports.ButtonStyle = {
    PRIMARY: 1,
    BLURPLE: 1,
    SECONDARY: 2,
    GREY: 2,
    SUCCESS: 3,
    GREEN: 3,
    DANGER: 4,
    RED: 4,
    LINK: 5,
    URL: 5,
};
exports.Interactions = {
    button: (options = {}) => {
        var _a, _b, _c, _d, _e, _f, _g;
        if (typeof options.style === "string") {
            const style = exports.ButtonStyle[options.style];
            if (style) {
                options.style = style;
            }
            else {
                options.style = 2;
            }
        }
        if (typeof options.url === "string" &&
            options.url.match(/(https?|discord):\/\//gi)) {
            options.style = 5;
            options.id = "";
        }
        return {
            custom_id: (_b = (_a = options === null || options === void 0 ? void 0 : options.id) !== null && _a !== void 0 ? _a : options === null || options === void 0 ? void 0 : options.custom_id) !== null && _b !== void 0 ? _b : "",
            label: (_d = (_c = options === null || options === void 0 ? void 0 : options.title) !== null && _c !== void 0 ? _c : options === null || options === void 0 ? void 0 : options.label) !== null && _d !== void 0 ? _d : "",
            type: (_e = options === null || options === void 0 ? void 0 : options.type) !== null && _e !== void 0 ? _e : 2,
            style: (options === null || options === void 0 ? void 0 : options.style) ? options.style : 2,
            disabled: typeof options.disabled === "boolean"
                ? options.disabled
                : false,
            emoji: (_f = options === null || options === void 0 ? void 0 : options.emoji) !== null && _f !== void 0 ? _f : undefined,
            url: (_g = options === null || options === void 0 ? void 0 : options.url) !== null && _g !== void 0 ? _g : undefined,
        };
    },
    select: (options = {}) => {
        if (options.type === v10_1.ComponentType.StringSelect &&
            !Array.isArray(options.options)) {
            throw new Error(`[Interactions#select]: The 'options' isn't an array.`);
        }
        const data = {
            custom_id: "",
            placeholder: "",
            min_values: 0,
            max_values: 0,
            options: [],
            type: 3,
            disabled: false,
        };
        if (utils_1.is.string(options.id)) {
            data.custom_id = options.id;
        }
        else if (utils_1.is.string(options.custom_id)) {
            data.custom_id = options.custom_id;
        }
        if (utils_1.is.string(options.placeholder)) {
            data.placeholder = options.placeholder;
        }
        else if (utils_1.is.string(options.holder)) {
            data.placeholder = options.holder;
        }
        if (utils_1.is.number(options.max_values)) {
            data.max_values = options.max_values;
        }
        else if (utils_1.is.number(options.max)) {
            data.max_values = options.max;
        }
        if (utils_1.is.number(options.min_values)) {
            data.min_values = options.min_values;
        }
        else if (utils_1.is.number(options.min)) {
            data.min_values = options.min;
        }
        if (utils_1.is.boolean(options.disabled)) {
            data.disabled = options.disabled;
        }
        if (utils_1.is.number(options.type)) {
            data.type = options.type;
        }
        if (utils_1.is.array(options.options)) {
            data.options = options.options;
        }
        return data;
    },
    modal: (options = { components: [] }) => {
        if (!utils_1.is.array(options.components)) {
            throw new Error(`[Interactions#modal]: 'components' isn't an array, or is empty.`);
        }
        const data = {
            custom_id: "",
            title: "",
            components: options.components,
        };
        if (utils_1.is.string(options.id)) {
            data.custom_id = options.id;
        }
        else if (utils_1.is.string(options.custom_id)) {
            data.custom_id = options.custom_id;
        }
        if (utils_1.is.string(options.title)) {
            data.title = options.title;
        }
        else if (utils_1.is.string(options.label)) {
            data.title = options.label;
        }
        return data;
    },
    textInput: (options = {}, row = true) => {
        const data = {
            type: 4,
            custom_id: undefined,
            style: 1,
            label: undefined,
            min_length: undefined,
            max_length: undefined,
            placeholder: undefined,
            value: undefined,
            required: false,
        };
        if (utils_1.is.number(options.type)) {
            data.type = options.type;
        }
        if (utils_1.is.string(options.id)) {
            data.custom_id = options.id;
        }
        else if (utils_1.is.string(options.custom_id)) {
            data.custom_id = options.custom_id;
        }
        if (utils_1.is.string(options.label)) {
            data.label = options.label;
        }
        else if (utils_1.is.string(options.title)) {
            data.label = options.title;
        }
        if (utils_1.is.string(options.placeholder)) {
            data.placeholder = options.placeholder;
        }
        else if (utils_1.is.string(options.holder)) {
            data.placeholder = options.holder;
        }
        if (utils_1.is.boolean(options.required)) {
            data.required = options.required;
        }
        if (utils_1.is.string(options.value)) {
            data.value = options.value;
        }
        if (utils_1.is.number(options.style)) {
            data.style = options.style;
        }
        if (utils_1.is.number(options.min)) {
            data.min_length = options.min;
        }
        else if (utils_1.is.number(options.min_length)) {
            data.min_length = options.min_length;
        }
        if (utils_1.is.number(options.max)) {
            data.max_length = options.max;
        }
        else if (utils_1.is.number(options.max_length)) {
            data.max_length = options.max_length;
        }
        if (row) {
            return { type: 1, components: [data] };
        }
        return data;
    },
};
exports.Duration = {
    parse: (value) => {
        const MATCHES_ALL = value.match(/\d+\s*[A-Za-z]+/g);
        if (MATCHES_ALL) {
            let totalTime = 0;
            for (const dur of MATCHES_ALL) {
                const n = dur.match(/\d+/g) || ["0"];
                const [num, [str]] = [
                    parseInt(n[0]),
                    dur.match(/[A-Za-z]+/g) || [""],
                ];
                if (isNaN(num)) {
                    totalTime = 0;
                }
                else {
                    totalTime += num * exports.Duration.determineTimeType(str);
                }
            }
            if (totalTime) {
                return totalTime;
            }
        }
        return null;
    },
    determineTimeType: (str) => {
        switch (str) {
            case "ms":
            case "millisecond":
            case "milliseconds":
                return 1;
            case "s":
            case "second":
            case "seconds":
                return 1000;
            case "m":
            case "min":
            case "mins":
            case "minute":
            case "minutes":
                return 60 * 1000;
            case "h":
            case "hr":
            case "hour":
            case "hours":
                return 60 * 60 * 1000;
            case "d":
            case "day":
            case "days":
                return 24 * 60 * 60 * 1000;
            case "w":
            case "week":
            case "weeks":
                return 7 * 24 * 60 * 60 * 1000;
            case "mo":
            case "month":
            case "months":
                return 30 * 24 * 60 * 60 * 1000;
            case "y":
            case "year":
            case "years":
                return 365 * 24 * 60 * 60 * 1000;
            default:
                return 1;
        }
    },
    validate: (value) => {
        const MATCHES_ALL = value.match(/\d+\s*[A-Za-z]+/g);
        if (MATCHES_ALL) {
            for (const match of MATCHES_ALL) {
                const [num, str] = [
                    match.match(/\d+/g),
                    match.match(/[A-Za-z]+/g),
                ];
                if (!num || num.length !== 1) {
                    return false;
                }
                if (!str || str.length !== 1) {
                    return false;
                }
                if (!Number.isInteger(parseInt(num[0]))) {
                    return false;
                }
                if (!exports.Duration.timeIds.has(str[0])) {
                    return false;
                }
            }
            return true;
        }
        return false;
    },
    timeIds: new Set([
        "ms",
        "millisecond",
        "milliseconds",
        "s",
        "second",
        "seconds",
        "m",
        "min",
        "mins",
        "minute",
        "minutes",
        "h",
        "hr",
        "hrs",
        "hour",
        "hours",
        "d",
        "day",
        "days",
        "w",
        "week",
        "weeks",
        "mo",
        "month",
        "months",
        "y",
        "year",
        "years",
    ]),
};
exports.SlashBuilder = {
    TEXT_BASED_CHANNELS: [0, 5, 11, 12],
    types: {
        sub_command: 1,
        sub_group: 2,
        string: 3,
        integer: 4,
        boolean: 5,
        user: 6,
        channel: 7,
        role: 8,
        mentionable: 9,
        number: 10,
        context: {
            user: 2,
            message: 3,
        },
    },
    context: {
        user: (name, options) => exports.SlashBuilder.create(name, "", Object.assign({ type: exports.SlashBuilder.types.context.user }, options)),
        message: (name, options) => exports.SlashBuilder.create(name, "", Object.assign({ type: exports.SlashBuilder.types.context.message }, options)),
    },
    choice: (name, value, name_localizations) => {
        return { name, value, name_localizations };
    },
    option: (data) => {
        var _a, _b;
        // @ts-ignore
        const _data = Object.assign({}, data);
        if ((_a = data.locale) === null || _a === void 0 ? void 0 : _a.names) {
            _data.name_localizations = data.locale.names;
        }
        if ((_b = data.locale) === null || _b === void 0 ? void 0 : _b.descriptions) {
            _data.description_localizations = data.locale.descriptions;
        }
        if ("locale" in _data) {
            delete _data["locale"];
        }
        return _data;
    },
    create: (name, description, options = {}) => {
        var _a, _b;
        const obj = {
            name,
            description,
            name_localizations: undefined,
            description_localizations: undefined,
            options: [],
            type: undefined,
            dm_permission: false,
            default_member_permissions: undefined,
        };
        if ((_a = options === null || options === void 0 ? void 0 : options.locale) === null || _a === void 0 ? void 0 : _a.names) {
            obj.name_localizations = options.locale.names;
        }
        if ((_b = options === null || options === void 0 ? void 0 : options.locale) === null || _b === void 0 ? void 0 : _b.descriptions) {
            obj.description_localizations = options.locale.descriptions;
        }
        if (utils_1.is.array(options.options)) {
            obj["options"] = options.options;
        }
        if (utils_1.is.number(options.type)) {
            obj.type = options.type;
        }
        if ("dmPermission" in options && utils_1.is.boolean(options.dmPermission)) {
            obj["dm_permission"] = options.dmPermission;
        }
        else if ("dm_permission" in options &&
            utils_1.is.boolean(options.dm_permission)) {
            obj["dm_permission"] = options.dm_permission;
        }
        if ("default_member_permissions" in options &&
            utils_1.is.string(options.default_member_permissions)) {
            obj.default_member_permissions = options.default_member_permissions;
        }
        else if ("defaultMemberPermissions" in options &&
            utils_1.is.string(options.defaultMemberPermissions)) {
            obj.default_member_permissions = options.defaultMemberPermissions;
        }
        return obj;
    },
};
