// @elara-services/packages@6.1.0/dist/src/lib/minesweeper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Minesweeper = void 0;
class Minesweeper {
    constructor(opts) {
        var _a, _b, _c, _d;
        this.rows = (_a = opts === null || opts === void 0 ? void 0 : opts.rows) !== null && _a !== void 0 ? _a : 5;
        this.columns = (_b = opts === null || opts === void 0 ? void 0 : opts.columns) !== null && _b !== void 0 ? _b : 5;
        this.mines = (_c = opts === null || opts === void 0 ? void 0 : opts.mines) !== null && _c !== void 0 ? _c : 5;
        this.matrix = [];
        this.types = {
            mine: (_d = opts === null || opts === void 0 ? void 0 : opts.emote) !== null && _d !== void 0 ? _d : "bomb",
            numbers: [
                "zero",
                "one",
                "two",
                "three",
                "four",
                "five",
                "six",
                "seven",
                "eight",
            ],
        };
    }
    generateEmptyMatrix() {
        for (let i = 0; i < this.rows; i++) {
            const arr = new Array(this.columns).fill(this.types.numbers[0]);
            this.matrix.push(arr);
        }
    }
    plantMines() {
        for (let i = 0; i < this.mines; i++) {
            const x = Math.floor(Math.random() * this.rows);
            const y = Math.floor(Math.random() * this.columns);
            if (this.matrix[x][y] === this.types.mine) {
                i--;
            }
            else {
                this.matrix[x][y] = this.types.mine;
            }
        }
    }
    getNumberOfMines(x, y) {
        if (this.matrix[x][y] === this.types.mine) {
            return this.types.mine;
        }
        let counter = 0;
        const hasLeft = y > 0;
        const hasRight = y < this.columns - 1;
        const hasTop = x > 0;
        const hasBottom = x < this.rows - 1;
        counter += +(hasTop &&
            hasLeft &&
            this.matrix[x - 1][y - 1] === this.types.mine);
        counter += +(hasTop && this.matrix[x - 1][y] === this.types.mine);
        counter += +(hasTop &&
            hasRight &&
            this.matrix[x - 1][y + 1] === this.types.mine);
        counter += +(hasLeft && this.matrix[x][y - 1] === this.types.mine);
        counter += +(hasRight && this.matrix[x][y + 1] === this.types.mine);
        counter += +(hasBottom &&
            hasLeft &&
            this.matrix[x + 1][y - 1] === this.types.mine);
        counter += +(hasBottom && this.matrix[x + 1][y] === this.types.mine);
        counter += +(hasBottom &&
            hasRight &&
            this.matrix[x + 1][y + 1] === this.types.mine);
        return this.types.numbers[counter];
    }
    start() {
        if (this.rows * this.columns <= this.mines * 2) {
            return null;
        }
        this.generateEmptyMatrix();
        this.plantMines();
        this.matrix = this.matrix.map((row, x) => row.map((col, y) => this.getNumberOfMines(x, y)));
        return this.matrix;
    }
}
exports.Minesweeper = Minesweeper;
