// @elara-services/packages@6.1.0/dist/src/lib/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tasks = void 0;
const utils_1 = require("@elara-services/utils");
const node_schedule_1 = require("node-schedule");
class Tasks extends null {
    static create({ id, time, shouldCancel } = {
        id: "",
        time: "",
        shouldCancel: true,
    }, run) {
        if (!utils_1.is.string(id) || !utils_1.is.string(time) || !utils_1.is.boolean(shouldCancel)) {
            return `You failed to provide one of the following options: 'id', 'time' or 'shouldCancel'`;
        }
        if (node_schedule_1.scheduledJobs[id]) {
            return `Found (${id}) in the scheduled jobs, ignoring.`;
        }
        return (0, node_schedule_1.scheduleJob)(id, time, () => {
            run();
            if (shouldCancel) {
                return void (0, node_schedule_1.cancelJob)(id);
            }
        });
    }
    static delete(id) {
        return (0, node_schedule_1.cancelJob)(id);
    }
    static has(id) {
        return node_schedule_1.scheduledJobs[id] ? true : false;
    }
}
exports.Tasks = Tasks;
