// @elara-services/bluesky@1.0.4/dist/src/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatPost = exports.chunk = exports.checkIfNew = exports.getRandomKey = exports.transformText = exports.transformKeys = exports.base = exports.baseService = void 0;
const tslib_1 = require("tslib");
const utils_1 = require("@elara-services/utils");
const api_1 = require("@atproto/api");
tslib_1.__exportStar(require("./parser"), exports);
exports.baseService = "https://bsky.social";
exports.base = "https://bsky.app";
function transformKeys(keys) {
    if (!utils_1.is.array(keys)) {
        keys = [keys];
    }
    return keys;
}
exports.transformKeys = transformKeys;
function transformText(post) {
    const t = new api_1.RichText({
        text: post.record?.text ?? "",
        facets: post.record.facets || [],
    });
    let text = post.record?.text || "";
    if (utils_1.is.object(post.record)) {
        if (utils_1.is.array(post.record.facets)) {
            for (const f of post.record.facets) {
                const str = t.unicodeText.slice(f.index.byteStart, f.index.byteEnd);
                const c = f.features.find((c) => c.$type === "app.bsky.richtext.facet#link");
                if (c) {
                    text = text.replace(new RegExp(str, "i"), `[${str}](${c.uri})`);
                }
            }
        }
    }
    return text;
}
exports.transformText = transformText;
function getRandomKey(keys) {
    keys = transformKeys(keys);
    return keys[Math.floor(Math.random() * keys.length)];
}
exports.getRandomKey = getRandomKey;
function checkIfNew(post, seconds) {
    if (!post.record?.createdAt || !utils_1.is.number(seconds)) {
        return false;
    }
    if (Date.now() - new Date(post.record.createdAt).getTime() <=
        seconds + 1200) {
        return true;
    }
    return false;
}
exports.checkIfNew = checkIfNew;
function chunk(arr, size) {
    const array = [];
    for (let i = 0; i < arr.length; i += size) {
        array.push(arr.slice(i, i + size));
    }
    return array;
}
exports.chunk = chunk;
function formatPost(data) {
    const { post } = data;
    const isRepost = data.reason?.$type === "app.bsky.feed.defs#reasonRepost";
    const reason = data.reason;
    return {
        createdAt: post.record?.createdAt || "",
        text: post.record?.text || "",
        type: isRepost ? "repost" : "post",
        images: (post.embed?.images || []).map((c) => c.fullsize) || [],
        media: (post.embed?.media || []),
        author: {
            handle: post.author.handle,
            username: post.author.displayName || post.author.handle,
            id: post.author.did,
            avatar: post.author.avatar || null,
        },
        links: {
            url: `${exports.base}/profile/${post.author.handle}/post/${post.uri.split("app.bsky.feed.post/")[1]}`,
            uri: post.uri,
            cid: post.cid,
        },
        counts: {
            likes: post.likeCount || 0,
            replies: post.replyCount || 0,
            reposts: post.repostCount || 0,
        },
        reposted: isRepost && reason && reason.by
            ? {
                createdAt: reason.indexedAt,
                user: {
                    username: reason.by.displayName || "",
                    handle: reason.by.handle || "",
                    avatar: reason.by.avatar || "",
                    id: reason.by.did || "",
                },
            }
            : null,
    };
}
exports.formatPost = formatPost;
