// @elara-services/bluesky@1.0.4/dist/src/utils/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = exports.p = void 0;
exports.p = {
    repost: {
        user: {
            name: "%repost.user.username%",
            handle: "%repost.user.handle%",
            id: "%repost.user.id%",
            avatar: "%repost.user.avatar%",
        },
        createdAt: "%repost.created_at%",
    },
    createdAt: "%created_at%",
    text: "%text%",
    type: "%type%",
    replies: "%replies%",
    likes: "%likes%",
    reposts: "%reposts%",
    user: {
        name: "%user.username%",
        handle: "%user.handle%",
        id: "%user.id%",
        avatar: "%user.avatar%",
    },
    links: {
        url: "%links.url%",
        uri: "%links.uri%",
        cid: "%links.cid%",
    },
    images: (num) => `%images${num ? `.[${num}]` : ""}%`,
};
function parse(obj, info) {
    let o = JSON.stringify(obj);
    const r = (name, str) => {
        o = o.replace(new RegExp(name, "gi"), str);
    };
    r(exports.p.user.handle, info.author.handle);
    r(exports.p.user.id, info.author.id);
    r(exports.p.user.avatar, info.author.avatar || "");
    r(exports.p.user.name, info.author.username);
    r(exports.p.text, info.text || "");
    r(exports.p.createdAt, info.createdAt);
    r(exports.p.type, info.type);
    r(exports.p.likes, info.counts.likes || 0);
    r(exports.p.replies, info.counts.replies || 0);
    r(exports.p.reposts, info.counts.reposts || 0);
    r(exports.p.links.url, info.links.url);
    r(exports.p.links.uri, info.links.uri);
    r(exports.p.links.cid, info.links.cid);
    if (info.type === "repost" && info.reposted) {
        r(exports.p.repost.createdAt, info.reposted.createdAt);
        r(exports.p.repost.user.name, info.reposted.user.username);
        r(exports.p.repost.user.handle, info.reposted.user.handle);
        r(exports.p.repost.user.avatar, info.reposted.user.avatar);
        r(exports.p.repost.user.id, info.reposted.user.id);
    }
    else {
        r(exports.p.repost.createdAt, "");
        r(exports.p.repost.user.id, "");
        r(exports.p.repost.user.handle, "");
        r(exports.p.repost.user.name, "");
        r(exports.p.repost.user.avatar, "");
    }
    if (info.images.length) {
        r(exports.p.images(), info.images.join(" "));
        for (let i = 0; i < info.images.length; i++) {
            // eslint-disable-next-line no-useless-escape
            r(exports.p.images(i), info.images[i] || "");
        }
    }
    else {
        r(exports.p.images(), "");
        for (let i = 0; i < 50; i++) {
            // eslint-disable-next-line no-useless-escape
            r(exports.p.images(i), "");
        }
    }
    return JSON.parse(o);
}
exports.parse = parse;
