// @elara-services/fetch@2.0.2/dist/Request.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Request = void 0;
/* eslint-disable no-prototype-builtins */
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const path_1 = require("path");
const querystring_1 = require("querystring");
const zlib_1 = require("zlib");
const Response_1 = require("./Response");
const supportedCompressions = ["gzip", "deflate"];
class Request {
    constructor(url, method = "GET") {
        this.url = typeof url === "string" ? new URL(url) : url;
        this.method = method;
        this.data = null;
        this.sendDataAs = null;
        this.reqHeaders = {};
        this.streamEnabled = false;
        this.compressionEnabled = false;
        this.timeoutTime = null;
        this.coreOptions = {};
        this.resOptions = {
            maxBuffer: 50 * 1000000, // 50 MB
        };
        return this;
    }
    query(a1, a2) {
        if (typeof a1 === "object") {
            Object.keys(a1).map((c) => this.url.searchParams.append(c, a1[c]));
        }
        else if (a1 && a2) {
            this.url.searchParams.append(a1, a2);
        }
        return this;
    }
    path(relativePath) {
        this.url.pathname = (0, path_1.join)(this.url.pathname, relativePath);
        return this;
    }
    body(data, sendAs) {
        this.sendDataAs =
            typeof data === "object" && !sendAs && !Buffer.isBuffer(data)
                ? "json"
                : sendAs
                    ? sendAs.toLowerCase()
                    : "buffer";
        this.data =
            this.sendDataAs === "form"
                ? // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    (0, querystring_1.stringify)(data)
                : this.sendDataAs === "json"
                    ? JSON.stringify(data)
                    : data;
        return this;
    }
    header(a1, a2) {
        if (typeof a1 === "object") {
            for (const name of Object.keys(a1)) {
                this.reqHeaders[name.toLowerCase()] =
                    a1[name];
            }
        }
        else if (a1 && a2) {
            this.reqHeaders[a1.toLowerCase()] = a2;
        }
        return this;
    }
    timeout(timeout) {
        this.timeoutTime = timeout;
        return this;
    }
    option(name, value) {
        this.coreOptions[name] = value;
        return this;
    }
    stream() {
        this.streamEnabled = true;
        return this;
    }
    compress() {
        this.compressionEnabled = true;
        if (!this.reqHeaders["accept-encoding"]) {
            this.reqHeaders["accept-encoding"] =
                supportedCompressions.join(", ");
        }
        return this;
    }
    send() {
        return new Promise((resolve, reject) => {
            if (this.data) {
                if (!this.reqHeaders.hasOwnProperty("content-type")) {
                    if (this.sendDataAs === "json") {
                        this.reqHeaders["content-type"] = "application/json";
                    }
                    else if (this.sendDataAs === "form") {
                        this.reqHeaders["content-type"] =
                            "application/x-www-form-urlencoded";
                    }
                }
                if (!this.reqHeaders.hasOwnProperty("content-length")) {
                    this.reqHeaders["content-length"] = Buffer.byteLength(this.data);
                }
            }
            const options = Object.assign({
                protocol: this.url.protocol,
                host: this.url.hostname,
                port: this.url.port,
                path: this.url.pathname +
                    (this.url.search === null ? "" : this.url.search),
                method: this.method,
                headers: this.reqHeaders,
            }, this.coreOptions);
            let req;
            const resHandler = (res) => {
                let stream = res;
                let Res;
                if (this.compressionEnabled) {
                    if (res.headers["content-encoding"] === "gzip") {
                        stream = res.pipe((0, zlib_1.createGunzip)());
                    }
                    else if (res.headers["content-encoding"] === "deflate") {
                        stream = res.pipe((0, zlib_1.createInflate)());
                    }
                }
                if (this.streamEnabled) {
                    resolve(stream);
                }
                else {
                    Res = new Response_1.Response(res, this.resOptions);
                    stream.on("error", (err) => reject(err));
                    stream.on("aborted", () => reject(new Error("Server aborted request")));
                    stream.on("data", (chunk) => {
                        Res._addChunk(chunk);
                        if (this.resOptions.maxBuffer !== null && Res.body) {
                            if (Res.body.length >
                                this.resOptions.maxBuffer) {
                                stream.destroy();
                                reject(`Received a response which was longer than acceptable when buffering. (${this.body.length} bytes)`);
                            }
                        }
                    });
                    stream.on("end", () => resolve(Res));
                }
            };
            if (this.url.protocol === "http:") {
                req = http_1.default.request(options, resHandler);
            }
            else if (this.url.protocol === "https:") {
                req = https_1.default.request(options, resHandler);
            }
            else {
                throw new Error(`Bad URL protocol: ${this.url.protocol}`);
            }
            if (this.timeoutTime) {
                req.setTimeout(this.timeoutTime, () => {
                    req.abort();
                    if (!this.streamEnabled)
                        reject(new Error("Timeout reached"));
                });
            }
            req.on("error", (err) => reject(err));
            if (this.data) {
                req.write(this.data);
            }
            req.end();
        });
    }
}
exports.Request = Request;
